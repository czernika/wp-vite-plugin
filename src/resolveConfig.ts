import { UserConfig } from 'vite'
import merge from 'merge'

const resolveConfig = (...configs: UserConfig[]): UserConfig => merge.recursive(...configs)

export default resolveConfig
