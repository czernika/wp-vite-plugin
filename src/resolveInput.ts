import type Config from '../types/Config'
import type Input from '../types/Input'
import * as path from 'path'
import ConfigInterface from '../interfaces/ConfigInterface'
import getThemePath from './getThemePath'

const resolveInput = (config: Config): Input => {
    let input: Input = {}

    /**
     * User passed single file as input
     *
     * @example
     * wolat('resources/js/app.js')
     */
    if (typeof config === 'string') {
        input = mapInput(config)
    }

    /**
     * User passed multiple files
     *
     * @example
     * wolat(['resources/js/app.js'])
     */
    if (Array.isArray(config)) {
        input = config.map((entry) => mapInput(entry))
    }

    /**
     * User passed input Config object
     *
     * @example
     * wolat({input: {app: 'resources/js/app.js'}})
     */
    if (typeof config === 'object' && !Array.isArray(config)) {
        const configInput = (config as ConfigInterface).input
        Object.keys(configInput).forEach(key => {
            input[key] = mapInput(configInput[key])
        })
    }

    return input
}

/**
 * Resolve entrypoint path based on theme path
 */
const mapInput = (input: string): string => path.resolve(getThemePath(), input)

export default resolveInput
