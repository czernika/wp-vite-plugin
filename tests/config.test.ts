import { it, expect, vi } from 'vitest'
import { UserConfig } from 'vite'
import resolveConfig from '../src/resolveConfig'
import resolveInput from '../src/resolveInput'

vi.mock('path', () => {
    return {
        resolve: (...args: string[]): string => ['/path/to', ...args].join('/')
    }
})

it('asserts configs were merged', () => {
    const pluginConfig: UserConfig = {
        build: {
            rollupOptions: {
                input: 'resources/common/js',
                output: {
                    entryFileNames: 'js/[name].[hash].js',
                }
            }
        }
    }

    const userConfig: UserConfig = {
        build: {
            rollupOptions: {
                input: 'resources/app/js',
                output: {
                    chunkFileNames: 'js/chunks/[name].[hash].js',
                }
            }
        }
    }

    const resolved = resolveConfig(pluginConfig, userConfig)

    expect(resolved).toMatchSnapshot()
})

it('asserts theme root resolves full path', () => {
    expect(resolveInput({
        input: 'resources/js/app.js',
        theme: 'wp-content/themes/test'
    })).toMatchObject('/path/to/wp-content/themes/test/resources/js/app.js')
})
