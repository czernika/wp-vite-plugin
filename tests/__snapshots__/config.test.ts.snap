// Vitest Snapshot v1

exports[`asserts configs were merged 1`] = `
{
  "build": {
    "rollupOptions": {
      "input": "resources/app/js",
      "output": {
        "chunkFileNames": "js/chunks/[name].[hash].js",
        "entryFileNames": "js/[name].[hash].js",
      },
    },
  },
}
`;
